### PR TITLE
Test Beam Creation Worker Instance

### PR DESCRIPTION
**Pull request by StevenGreen1**
Restructured test beam particle creation algorithm to work within worker instance.

### CODE REVIEW

File: larpandoracontent/LArThreeDReco/LArEventBuilding/TestBeamParticleCreationAlgorithm.cc [hunk 1]
```diff
@@ -84,21 +84,23 @@ StatusCode TestBeamParticleCreationAlgorithm::Run()
 
         PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::ParticleFlowObject::AlterMetadata(*this, pPrimaryPfo, pfoMetadata));
 
+        const std::string previousListName(LArPfoHelper::IsTrack(pPrimaryPfo) ? m_trackPfoListName : m_showerPfoListName);
```
Comment (PandoraPFA) [inline larpandoracontent/LArThreeDReco/LArEventBuilding/TestBeamParticleCreationAlgorithm.cc]: Take a const reference.

File: larpandoracontent/LArThreeDReco/LArEventBuilding/TestBeamParticleCreationAlgorithm.cc [hunk 2]
```diff
@@ -84,21 +84,23 @@ StatusCode TestBeamParticleCreationAlgorithm::Run()
...
+        PfoList pfoList;
+        pfoList.push_back(pPrimaryPfo);
```
Comment (PandoraPFA) [inline larpandoracontent/LArThreeDReco/LArEventBuilding/TestBeamParticleCreationAlgorithm.cc]: Can use e.g. `const PfoList pfoList(1, pPrimaryPfo);`

File: larpandoracontent/LArThreeDReco/LArEventBuilding/TestBeamParticleCreationAlgorithm.cc [hunk 3]
```diff
@@ -110,6 +112,12 @@ StatusCode TestBeamParticleCreationAlgorithm::Run()
                 PANDORA_THROW_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::SaveList<Vertex>(*this, m_vertexListName));
                 PANDORA_THROW_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::ReplaceCurrentList<Vertex>(*this, m_vertexListName));
             }
+            else
+            {
+                const Vertex *const pVertex(LArPfoHelper::GetVertex(pPfo));
+                PANDORA_THROW_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::RemoveFromPfo(*this, pPfo, pVertex));
+                PANDORA_THROW_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::AddToPfo(*this, pPrimaryPfo, pVertex));
```
Comment (PandoraPFA) [inline larpandoracontent/LArThreeDReco/LArEventBuilding/TestBeamParticleCreationAlgorithm.cc]: This one then ought to move into the NeutrinoVertices3D list, to match behaviour in other modes

File: larpandoracontent/LArThreeDReco/LArEventBuilding/TestBeamParticleCreationAlgorithm.cc [hunk 4]
```diff
@@ -110,6 +112,12 @@ StatusCode TestBeamParticleCreationAlgorithm::Run()
                 PANDORA_THROW_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::SaveList<Vertex>(*this, m_vertexListName));
                 PANDORA_THROW_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::ReplaceCurrentList<Vertex>(*this, m_vertexListName));
```
Comment (PandoraPFA) [inline larpandoracontent/LArThreeDReco/LArEventBuilding/TestBeamParticleCreationAlgorithm.cc]: This output list ought to be the NeutrinoVertices3D list to match behaviour for neutrinos.

File: larpandoracontent/LArThreeDReco/LArEventBuilding/TestBeamParticleCreationAlgorithm.cc [hunk 5]
```diff
@@ -130,15 +138,17 @@ StatusCode TestBeamParticleCreationAlgorithm::Run()
 StatusCode TestBeamParticleCreationAlgorithm::ReadSettings(const TiXmlHandle xmlHandle)
 {
     PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, XmlHelper::ReadValue(xmlHandle, "PfoListName", m_pfoListName));
```
Comment (PandoraPFA) [inline larpandoracontent/LArThreeDReco/LArEventBuilding/TestBeamParticleCreationAlgorithm.cc]: Update name for this PfoList to make it more self describing - ParentPfoList?

File: larpandoracontent/LArThreeDReco/LArEventBuilding/TestBeamParticleCreationAlgorithm.cc [hunk 6]
```diff
@@ -130,15 +138,17 @@ StatusCode TestBeamParticleCreationAlgorithm::Run()
...
 
     PANDORA_RETURN_RESULT_IF_AND_IF(STATUS_CODE_SUCCESS, STATUS_CODE_NOT_FOUND, !=, XmlHelper::ReadValue(xmlHandle,
         "KeepInteractionVertex", m_keepInteractionVertex));
 
     PANDORA_RETURN_RESULT_IF_AND_IF(STATUS_CODE_SUCCESS, STATUS_CODE_NOT_FOUND, !=, XmlHelper::ReadValue(xmlHandle,
         "KeepStartVertex", m_keepStartVertex));
 
-    if (m_keepStartVertex)
-        PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, XmlHelper::ReadValue(xmlHandle, "VertexListName", m_vertexListName));
+    PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, XmlHelper::ReadValue(xmlHandle, "VertexListName", m_vertexListName));
```
Comment (PandoraPFA) [inline larpandoracontent/LArThreeDReco/LArEventBuilding/TestBeamParticleCreationAlgorithm.cc]: ParentVertexListName?

**Comment by PandoraPFA** [comment]
Refactored and pushed to branch [feature/TestBeamParticleCreationWorker](https://github.com/PandoraPFA/LArContent/tree/feature/TestBeamParticleCreationWorker)

Please take a look and comment. I don't think there are any remaining issues, but testing would be appreciated.

Placing the algorithm at the end of the worker instance means it has to deal with *every* slice, whereas before it only ran over slices that were deemed worthy of being a test beam particle. This introduced threshold issues, with noisy warnings printed to screen, but all should now be OK!

Thanks!

**Comment by StevenGreen1** [comment]
I've looked at the new feature branch and it looks good now.  Much cleaner too.  I will run some test jobs now to make sure everything behaves as expected and get back to you with the results, but I imagine it's all good.

**Comment by PandoraPFA** [comment]
Merged.

**Pull request closed without merging**